distributed/scheduler.py

```diff
@@ -8205,6 +8280,127 @@ def _worker_full(ws: WorkerState, saturation_factor: float) -> bool:
+
+
+def family(
+    ts: TaskState, maxsize: int, widely_shared_cutoff: int
+) -> tuple[set[TaskState], set[TaskState]] | None:
+    """
+    All tasks in a family must be in memory at once to compute at least one common dependency.
+
+    Returns these ``sibling`` tasks, and the set of ``downstream`` (dependent) tasks that
+    the siblings will used to compute.
+
+    All ``siblings`` and ``downstream`` should be scheduled onto the same worker.
+    All ``siblings`` should be be scheduled at once---there's no benefit to queuing them.
+
+    For the purposes of identifying families:
+
+    * Linear chains are collapsed (traversed up and down)
+    * Widely-shared tasks (tasks with > ``widely_shared_cutoff`` dependents) are ignored
+
+    If the task's family is too large, or empty, returns None.
+    That is, if ``siblings`` would be larger than ``maxsize``, or ``downstream`` would be
+    larger than ``widely_shared_cutoff``, or ``ts.dependents`` is empty, returns None.
+    """
+    # TODO potentially could be useful to distinguish between 'too big'
+    # and empty family---we might want to schedule them differently.
+    # That is, maybe `None` and `(set(), set())` might not be synonymous.
+    if not ts.dependents or len(ts.dependents) > widely_shared_cutoff:
+        return None
+
+    siblings: set[TaskState] = set()
+    downstream: set[TaskState] = set()
+    # TODO maintain `seen` set to avoid repeated traversal? Should we add on the way down, or just back up?
+    for dts in ts.dependents:  # TODO even support multiple dependents?
+        # Traverse down linear chains
+        while ndts := _next_in_linear_chain(dts, widely_shared_cutoff):
+            # TODO check seen
+            dts = ndts
+
+        if dts in downstream:
+            # No need to traverse from a task we've already seen
+            continue
+
+        if dts in siblings:
+            # `siblings` and `downstream` are exclusive, and `siblings` takes priority
+            siblings.remove(dts)
+            continue
+
+        sibs = dts.dependencies
+        if len(sibs) == 1:
+            # Faster path: this is just a linear chain, so no siblings besides `ts`
+            downstream.add(dts)
+            continue
+
+        if len(sibs) < maxsize:
+            downstream.add(dts)
+            for sts in sibs:
+                # Traverse linear chains _back_ to root tasks
+                while (
+                    sts is not ts
+                    and sts not in downstream
+                    and (psts := _previous_in_linear_chain(sts, widely_shared_cutoff))
+                ):
+                    # TODO check seen
+                    sts = psts
+
+                if (
+                    sts is ts
+                    or len(sts.dependents)
+                    >= widely_shared_cutoff  # ignore widely-shared siblings
+                    or sts in downstream  # a->b, b->c, a->c. downstream takes priority.
+                ):
+                    continue
```

Comment:
IIUC this is the only place where widely_shared_cutoff should be interpreted as a cluster size. if so, we should easily be able to split this dynamic component off

Reply:
It's also important when walking chains up and down, as mentioned in other comments, for cases like https://github.com/dask/distributed/blob/0922e5c293474e89acac16623fd1741013e30453/distributed/tests/test_scheduler_family.py#L179-L184